File: kompas_maturalny_backend/app/graphql/mutations/users/sign_up_user.rb
module Mutations
  module Users
    class SignUpUser < BaseMutation
      argument :username, String, required: true
      argument :email, String, required: true
      argument :password, String, required: true
      argument :role, Types::Enums::Role, required: false

      field :user, Types::UserType, null: true
      field :errors, [String], null: false

      def resolve(username:nil, email: nil, password: nil, role: "schoolboy")
        user = User.create!(
            username: username,
            email: email,
            password: password,
            role: role
        )

        context[:current_user] = user

          {
              user: user,
              errors: []
          }

      rescue ActiveRecord::RecordInvalid => invalid
        GraphQL::ExecutionError.new(
            "Invalid Attributes for #{invalid.record.class.name}: " \
        "#{invalid.record.errors.full_messages.join(', ')}"
        )
      end
    end
  end
end

File: kompas_maturalny_web/graphql/mutations/sign_up_user.js
import gql from "graphql-tag";

export const SIGN_UP_USER = gql`
    mutation signUpUser($username: String!, $email: String!, $password: String!){
        signUpUser(input: {
            username: $username,
            email: $email,
            password: $password
        }
        ) {
            user {
                id
                email
                role
                authenticationToken
            }
        }
    }
`;